clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -244,6 +258,53 @@ public void forceClose() {
         initiateClose();
     }
 
+    private void getPID(long remainingTime) {
+        // Send an InitPIDRequest, wait for the response, retrieve the PID from the response if success. Return it.
+        // Otherwise throw exception.
+        // Pick a random broker to send the initPIDRequest to.
+        metadata.requestUpdate();
+        long start = time.milliseconds();
+        Node node = client.leastLoadedNode(time.milliseconds());
+        while (!client.ready(node, time.milliseconds()) && 0 < remainingTime) {
+            if (client.connectionFailed(node)) {
+                node = client.leastLoadedNode(time.milliseconds());
+                client.ready(node, time.milliseconds());
+            }
+            client.poll(100, time.milliseconds());
+            remainingTime = remainingTime - (time.milliseconds() - start);
+        }
+        if (remainingTime <= 0) {
+            return;
+        }
+
+        String nodeId = node.idString();
+        InitPidRequest.Builder builder = new InitPidRequest.Builder(null);
+        RequestCompletionHandler callback = new RequestCompletionHandler() {
+            public void onComplete(ClientResponse response) {
+                handleInitPidResponse(response);
+            }
+        };
+
+        ClientRequest request = client.newClientRequest(nodeId, builder, time.milliseconds(), true, callback);
+        client.send(request, time.milliseconds());
+        start = time.milliseconds();
+        while (!transactionState.hasPid() && 0 < remainingTime) {
```

Comment:
If the request fails immediately (say because of a disconnect), it seems like we're still stuck in this loop until expiration of the timeout. I think we need a way to break out if the request fails.